### ISSUE TITLE
Cannot compile on Mac OS X due to BoringSSL

### ISSUE THREAD

**Issue by ortutay** [tensorflow/tensorflow#12123]
------------------------

### System information

```
@:~/projects/tensorflow <master>$ cat tf_env.txt

== cat /etc/issue ===============================================
Darwin mn-mortutay 16.5.0 Darwin Kernel Version 16.5.0: Fri Mar  3 16:52:33 PST 2017; root:xnu-3789.51.2~3/RELEASE_X86_64 x86_64
Mac OS X 10.12.4

== are we in docker =============================================
No

== compiler =====================================================
Apple LLVM version 8.1.0 (clang-802.0.42)
Target: x86_64-apple-darwin16.5.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin

== uname -a =====================================================
Darwin mn-mortutay 16.5.0 Darwin Kernel Version 16.5.0: Fri Mar  3 16:52:33 PST 2017; root:xnu-3789.51.2~3/RELEASE_X86_64 x86_64

== check pips ===================================================
numpy (1.13.1)
protobuf (3.3.0)
tensorflow (1.2.1)

== check for virtualenv =========================================
False

== tensorflow import ============================================
tf.VERSION = 1.2.1
tf.GIT_VERSION = v1.2.0-5-g435cdfc
tf.COMPILER_VERSION = v1.2.0-5-g435cdfc
Sanity check: array([1], dtype=int32)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "tensorflow/__init__.py", line 24, in <module>
    from tensorflow.python import *
  File "tensorflow/python/__init__.py", line 49, in <module>
    from tensorflow.python import pywrap_tensorflow
  File "tensorflow/python/pywrap_tensorflow.py", line 52, in <module>
    raise ImportError(msg)
ImportError: Traceback (most recent call last):
  File "tensorflow/python/pywrap_tensorflow.py", line 41, in <module>
    from tensorflow.python.pywrap_tensorflow_internal import *
ImportError: No module named pywrap_tensorflow_internal


Failed to load the native TensorFlow runtime.

See https://www.tensorflow.org/install/install_sources#common_installation_problems

for some common reasons and solutions.  Include the entire stack trace
above this error message when asking for help.

== env ==========================================================
LD_LIBRARY_PATH is unset
DYLD_LIBRARY_PATH is unset

== nvidia-smi ===================================================
tf_env_collect.sh: line 105: nvidia-smi: command not found

== cuda libs  ===================================================
```

Building from commit:
```
* d74f65bac (Yun Peng, 3 hours ago) Make Windows Bazel GPU build work again (#11901)
```

### Describe the problem
I'm following the instructions for Mac source installation (https://www.tensorflow.org/install/install_sources, "Prepare environment for Mac OS" and then "Build the pip package") but I'm running into issues with BoringSSL.

It looks like the `-Wunused-but-set-parameter` and `-Wno-free-nonheap-object` flags are causing the compilation to fail, since `clang` does not support them. 

Expected behavior: The build system should detect this and handle it correctly, eg. by not using the un-available flags

Actual behavior: Error / fails to build

### Source code / logs

```
@:~/projects/tensorflow <master>$ bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package
WARNING: /Users/mortutay/projects/tensorflow/tensorflow/contrib/learn/BUILD:15:1: in py_library rule //tensorflow/contrib/learn:learn: target '//tensorflow/contrib/learn:learn' depends on deprecated target '//tensorflow/contrib/session_bundle:exporter': No longer supported. Switch to SavedModel immediately.
WARNING: /Users/mortutay/projects/tensorflow/tensorflow/contrib/learn/BUILD:15:1: in py_library rule //tensorflow/contrib/learn:learn: target '//tensorflow/contrib/learn:learn' depends on deprecated target '//tensorflow/contrib/session_bundle:gc': No longer supported. Switch to SavedModel immediately.
INFO: Found 1 target...
INFO: From Compiling external/protobuf_archive/python/google/protobuf/internal/api_implementation.cc:
warning: unknown warning option '-Wunused-but-set-parameter'; did you mean '-Wunused-parameter'? [-Wunknown-warning-option]
warning: unknown warning option '-Wno-free-nonheap-object'; did you mean '-Wno-sequence-point'? [-Wunknown-warning-option]
2 warnings generated.
INFO: From Compiling external/swig/Source/Swig/typemap.c [for host]:
warning: unknown warning option '-Wunused-but-set-parameter'; did you mean '-Wunused-parameter'? [-Wunknown-warning-option]
warning: unknown warning option '-Wno-free-nonheap-object'; did you mean '-Wno-sequence-point'? [-Wunknown-warning-option]
2 warnings generated.
INFO: From Compiling external/grpc/src/core/lib/debug/trace.c:
warning: unknown warning option '-Wunused-but-set-parameter'; did you mean '-Wunused-parameter'? [-Wunknown-warning-option]
warning: unknown warning option '-Wno-free-nonheap-object'; did you mean '-Wno-sequence-point'? [-Wunknown-warning-option]
2 warnings generated.
INFO: From Compiling external/grpc/src/cpp/codegen/codegen_init.cc:
warning: unknown warning option '-Wunused-but-set-parameter'; did you mean '-Wunused-parameter'? [-Wunknown-warning-option]
warning: unknown warning option '-Wno-free-nonheap-object'; did you mean '-Wno-sequence-point'? [-Wunknown-warning-option]
2 warnings generated.
INFO: From Compiling external/grpc/src/core/ext/transport/chttp2/alpn/alpn.c:
warning: unknown warning option '-Wunused-but-set-parameter'; did you mean '-Wunused-parameter'? [-Wunknown-warning-option]
warning: unknown warning option '-Wno-free-nonheap-object'; did you mean '-Wno-sequence-point'? [-Wunknown-warning-option]
2 warnings generated.
ERROR: /private/var/tmp/_bazel_mortutay/dacb21c644505cd819865fa365d2b69e/external/boringssl/BUILD:116:1: C++ compilation of rule '@boringssl//:crypto' failed (Exit 1).
error: unknown warning option '-Wunused-but-set-parameter'; did you mean '-Wunused-parameter'? [-Werror,-Wunknown-warning-option]
error: unknown warning option '-Wno-free-nonheap-object'; did you mean '-Wno-sequence-point'? [-Werror,-Wunknown-warning-option]
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 2.829s, Critical Path: 0.91s
```

**Comment by mountaintom** [comment]
Hi,

I recently had issues compiling BoringSSL because the -Werror complier flag was set in the BoringSSL build. If you error log is complete, it looks like the same warnings that were tolerated in building other parts of the code, were treated as fatal errors in the BoringSSL build. 

If the -Werror flag is your problem if would be best to find where it is set. However, I used a brute force solution to get the job done.

```bash
find . -type f -not -regex ".*\.\(o\|h\|c\|a\|t\|cc\|so\)" -exec sed -i -e "s/-Werror/-Wno-excessive-errors/g" {} \;
```

Hope this helps.

**Comment by tatatodd** [comment]
@mountaintom Thanks for the analysis!  Your explanation looks spot-on to me.

@av8ramit Do you know who might be a good candidate to follow-up on Mac builds?

**Comment by lissahyacinth** [comment]
Experienced the same issue with boringSSL, dug into it from and found Werror is being set in [add_boringssl_s390x.patch](https://github.com/tensorflow/tensorflow/blob/master/third_party/boringssl/add_boringssl_s390x.patch) on line 71, which is causing the build failure.

```+boringssl_copts = select({
+    ":windows": [
+        "-DWIN32_LEAN_AND_MEAN",
+    ],
+    "//conditions:default": [
+        # Assembler option --noexecstack adds .note.GNU-stack to each object to
+        # ensure that binaries can be built with non-executable stack.
+        "-Wa,--noexecstack",
+
+        # This is needed on Linux systems (at least) to get rwlock in pthread.
+        "-D_XOPEN_SOURCE=700",
+
+        # This list of warnings should match those in the top-level CMakeLists.txt.
+        "-Wall",
+        "-Werror",
+        "-Wformat=2",
+        "-Wsign-compare",
+        "-Wmissing-field-initializers",
+        "-Wwrite-strings",
+        "-Wshadow",
+        "-fno-common",
+    ],
+})
```

`-Werror` can just be replaced with `Wno-excessive-errors`, as suggested by @mountaintom , and TF will successfully build. Not sure if this is a sufficient fix as I'm not familiar with the library, but it does bypass the issue for now.

**Comment by luchensk** [comment]
+1, the issue still there when compiling from source in master branch, after used the way of @lissahyacinth , as below:

$ bazel build --config=opt //tensorflow/tools/pip_package:build_pip_package --verbose_failures
WARNING: /Users/sunkai/work/src/tensorflow/tensorflow/core/BUILD:1632:1: in includes attribute of cc_library rule //tensorflow/core:framework_headers_lib: '../../external/nsync/public' resolves to 'external/nsync/public' not below the relative path of its package 'tensorflow/core'. This will be an error in the future. Since this rule was created by the macro 'cc_header_only_library', the error might have been caused by the macro implementation in /Users/sunkai/work/src/tensorflow/tensorflow/core/BUILD:1632:1.
ERROR: /Users/sunkai/work/src/tensorflow/tensorflow/tools/pip_package/BUILD:100:1: no such package '@boringssl//': Traceback (most recent call last):
	File "/Users/sunkai/work/src/tensorflow/tensorflow/workspace.bzl", line 119
		_apply_patch(repo_ctx, repo_ctx.attr.patch_file)
	File "/Users/sunkai/work/src/tensorflow/tensorflow/workspace.bzl", line 110, in _apply_patch
		_execute_and_check_ret_code(repo_ctx, cmd)
	File "/Users/sunkai/work/src/tensorflow/tensorflow/workspace.bzl", line 94, in _execute_and_check_ret_code
		fail("Non-zero return code({1}) when ...))
Non-zero return code(2) when executing 'patch -p1 -d /private/var/tmp/_bazel_sunkai/f32adcfaaeaeed39ed3812ab214ba306/external/boringssl -i /Users/sunkai/work/src/tensorflow/third_party/boringssl/add_boringssl_s390x.patch':
Stdout: patching file src/include/openssl/base.h
patching file BUILD

Stderr: patch: **** malformed patch at line 88:      #"//conditions:default": ["-DOPENSSL_NO_ASM"],

 and referenced by '//tensorflow/tools/pip_package:licenses'.
ERROR: Analysis of target '//tensorflow/tools/pip_package:build_pip_package' failed; build aborted.
INFO: Elapsed time: 3.261s

Is there any idea? Thanks.

**Comment by luchensk** [comment]
Is there some update? Looking forward to the response, thanks.

**Comment by keithhans** [comment]
I don't have problem with the way @lissahyacinth provided.

Please take a look at this error:
Stderr: patch: **** malformed patch at line 88: #"//conditions:default": ["-DOPENSSL_NO_ASM"],
Looks like your patch file is not correct.

**Comment by RenatoUtsch** [comment]
This seems to be a bug with newer Bazel releases. I started having this with all my C++ projects that use Bazel. They all show these two warnings when compiling C++ files.

It looks like it adds these two flags when compiling with clang (I'm on Linux), even though it shouldn't.

@gunan marking you as you pointed these warnings on the duplicate bug.

**Comment by facaiy** [comment]
@quaeler Could you paste your the version of bazel which works well on mac #12266? As reported by @RenatoUtsch on bazelbuild/bazel#3622 , the warning seems to be brought by new version of bazel.

**Comment by quaeler** [comment]
Yep:
```
autogenic:tensorflow loki$ bazel version
Build label: 0.5.3-homebrew
Build target: bazel-out/darwin_x86_64-opt/bin/src/main/java/com/google/devtools/build/lib/bazel/BazelServer_deploy.jar
Build time: Sun Aug 6 12:54:10 2017 (1502024050)
Build timestamp: 1502024050
Build timestamp as int: 1502024050
autogenic:tensorflow loki$
```

**Comment by RenatoUtsch** [comment]
Interesting, Bazel does not add these flags on my Mac OS, while it adds them to the exact same rule in Linux. Maybe this is related to something else, and not only Bazel's version.

**Comment by RenatoUtsch** [comment]
Okay, upon digging a but more this does not seem to be related to bazel's version, but still seems to be a bazel bug. Follow bazelbuild/bazel#3622 if interested

**Comment by luchensk** [comment]
Thanks @RenatoUtsch . I have put my env info on [bazelbuild/bazel#3622](https://github.com/bazelbuild/bazel/issues/3622).

**Comment by luchensk** [comment]
@keithhans thanks for the point. 
But, I didn't change the line 88(should be 87 in original add_boringssl_s390x.patch) of `#"//conditions:default": ["-DOPENSSL_NO_ASM"]`.

**Comment by luchensk** [comment]
I find the reason why it didn't take effect by "replacing" `-Werror` with `-Wno-excessive-errors` since I wrongly added `#` to try to comment `-Werror` instead of replacing. Now it works and thanks.

**Comment by snnn** [comment]
Where does "-Wunused-but-set-parameter" come from? What's your bazel version? If it is not 0.5.4, could you do an upgrade?
Bazel shouldn't add this arg to a compiler that doesn't support it.

**Comment by RenatoUtsch** [comment]
Yeah, removing -Werror is just a hack and doesn't solve the real problem.

I could definitely reproduce bazel adding those two flags wrongly to clang
on 0.5.3. In a few hours I can update and try it again with 0.5.4.

**Comment by RenatoUtsch** [comment]
This is fixed with Bazel 0.5.4.

Just upgrade and these warnings won't happen anymore. No need for the `-Werror` hack too.

**Comment by gunan** [comment]
Closing based on the latest comments.
